lib/rubygems.rb

```diff
@@ -1077,11 +1096,10 @@ def self.load_plugin_files(plugins) # :nodoc:
   end
 
   ##
-  # Find the 'rubygems_plugin' files in the latest installed gems and load
-  # them
+  # Find rubygems plugin files in the standard location and load them
 
   def self.load_plugins
-    load_plugin_files find_latest_files('rubygems_plugin', false)
+    load_plugin_files Gem::Util.glob_files_in_dir("*#{Gem.plugin_suffix_pattern}", plugins_dir)
```

Comment:
This is the most important change in this PR. We go from:

* Look for requirable files named 'rubygems_plugin' in all installed gems, which means looking in every installed gem's folder, evaluate its gemspec to find its 'require_paths' (usually `lib/`) and actually look in that folder.

to: 

* Require any files in the standard rubygems plugins folder.

The rest of the code in this PR is essentially code to be able to do this particular change safely and in a backwards compatible way.

Reply:
Will this respect $LOAD_PATH? If I activate a gem, I'd expect its plugin to be added to `gem`, and if I deactivate it, I'd expect its plugin not to be added.

Reply:
What do you mean exactly by "I expect its plugin to be added to `gem`"? In any case, the end user behavior of `rubygems` should be unchanged by this PR.

Reply:
No problem! So the current way to load plugins is to scan for all active gems that have a `rubygems_require.rb` in the `lib` folder. So, if a gem isn't activated, its plugins aren't added, and when it is activated (by having its `lib` folder listed in the `$LOAD_PATH`), its plugins are added.

What I mean by "I expect its plugin to be added to `gem`" is just the first part, the normal, regular expected behaviour. I mentioned it to be symmetrical with the deactivation case.

I think, with this refactor, if I install a gem, its plugin will be added to `gem` regardless of if the gem is activated or not. With the pre-refactor implementation, gems that weren't activated didn't have their plugins added to `gem`, because activation meant "had their gem `lib` folder added to `$LOAD_PATH`", and that's where `Gem::Specification.latest_specs` looks.

Reply:
I'm pretty sure it doesn't work the way you think, although I agree it would probably make more sense the way you explain, but I didn't want to change the behaviour just now.

Rubygems loads plugins for the latest version of every installed gem (in addition to plugins present in the $LOAD_PATH), no matter whether it has yet been activated or not.

Reply:
As an example, if I have the following `Gemfile`s:
```
source 'https://rubygems.org'

gem 'gem-compiler'
```

```
source 'https://rubygems.org'
```

Then, when I run `bundle exec gem compile --help` with the first one, I'd expect to see the help docs for `gem-compiler`'s plugin, and for the second one I'd expect to see `Unknown command compile`.

Reply:
I see. Everything I said applies for a "non-bundled context".

But `bundler` further modifies `rubygems` plugins system to adapt everything `rubygems` does to apply to just the gems in the current bundle. So you're right, maybe whatever monkeypatches `bundler` is doing to adapt to `rubygems` plugins system need to be modified.

I'll have a look and get back to you.

Reply:
I just tried this and it seems to work as expected? Did you try it?

Reply:
I tried it, but I don't have any new-style plugin gems to install. With `gem-compile`, it works because of the call to [`load_env_plugins` in `Gem::GemRunner`](https://github.com/rubygems/rubygems/blob/b933fb6b103f7b8aae7881e5f80fbf5f9d3e26dc/lib/rubygems/gem_runner.rb#L16). Is there a gem you are using that uses the reworked plugin system?

Reply:
Also thanks for following up on this!

Reply:
No problem. What do you mean by "new style plugins"?

Reply:
The ones that register themselves by creating a file in `Gem.plugins_dir`.

Reply:
That is _all_ rubygems plugins actually. If your gem includes a `rubygems_plugin.rb` file, it will be registered like that in the next rubygems version. In other words, `gem-compiler` is an example of such a plugin, and it works as expected for me.

Reply:
ah, because the `load plugin` statement in `load_plugin_files` catches the exception when the `require` fails, because the gem is not in the `$LOAD_PATH`. So, the answer to my original question is 'yes, it does respect `$LOAD_PATH`'.

Will you be removing the call to `load_env_plugins` in the future?

Reply:
I don't think that rescue is actually involved here, since the `require` uses absolute paths, so it does not involve `$LOAD_PATH`. The reason this "works" is that even if the "compile" command is actually registered, it can never be loaded when running through `bundler` since the command file is not on the `$LOAD_PATH`.

See the error against current rubygems:

```
$ bundle exec gem compile
ERROR:  While executing gem ... (Gem::CommandLineError)
    Unknown command compile
``` 

vs the new error

```
$ bundle exec gem compile
ERROR:  Loading command: compile (LoadError)
	cannot load such file -- rubygems/commands/compile_command
ERROR:  While executing gem ... (NoMethodError)
    undefined method `invoke_with_build_args' for nil:NilClass
```

This means we _do_ need to adapt `bundler` to work with the new plugin system.

I'm working on a fix for this [here](https://github.com/rubygems/bundler/compare/catch_up_with_rubygems_plugin_loading_changes).

Reply:
Ok, so I spent a fair bit of time on this today. Turns out the different error messages are unrelated to this change in rubygems, but depend on the fact whether `gem-compiler` is installed globally (in addition to through the `Gemfile` or not).

Even if not caused by this particular change in rubygems, I think that is a bug since the global environment should never "pollute" a `bundle exec` context like that, so I'm fixing it in the branch I pointed to.